### PR TITLE
cupy build fix for SWDEV-475036

### PR DESCRIPTION
**Pull request by hongxiayang**
Kiran Thumma reached out to me regarding a build error he had: 
http://rocm-ci.amd.com/job/vllm-test/8/
The error is:
 sh: 1: gfx942: not found
05:50:57  failed to execute:/opt/rocm/lib/llvm/bin/clang++  --offload-arch=gfx90a;gfx942 -O3 --driver-mode=g++ -O3 --hip-link  -o "/tmp/tmpqall0p3c/test.out" -x hip /tmp/tmpqall0p3c/test.cu

This PR is to fix that build error.

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: Dockerfile.rocm [hunk 1]
```diff
@@ -93,7 +93,7 @@ RUN git clone  https://github.com/ROCm/cupy.git \
     && git submodule update --init --recursive \
     && pip install mpi4py-mpich scipy==1.9.3 cython==0.29.* \
     && CC=$MPI_HOME/bin/mpicc python -m pip install mpi4py \
-    && CUPY_INSTALL_USE_HIP=1 ROCM_HOME=/opt/rocm HCC_AMDGPU_TARGET=${PYTORCH_ROCM_ARCH} \
+    && CUPY_INSTALL_USE_HIP=1 ROCM_HOME=/opt/rocm HCC_AMDGPU_TARGET="gfx90a,gfx942,gfx1100" \
```
Comment (shajrawi) [inline Dockerfile.rocm:96]: can we have it as a variable that we can set instead of hardcoded on this line? Even if different than the one above?

**Review by shajrawi — COMMENTED** [review]
small nit

File: Dockerfile.rocm (hunk 1 above)
Comment (hongxiayang) [reply Dockerfile.rocm]: done

**Review by shajrawi — DISMISSED** [review]
LGTM! Thank you for doing this :)

**Comment by hongxiayang** [comment]
> LGTM! Thank you for doing this :)

np. Kiran is testing, and he will let me know the result before we merge it.

**Comment by hongxiayang** [comment]
context is that this is for rom6.3. 
https://compute-artifactory.amd.com:5000/artifactory/rocm-docker-experimental/rocm-plus-docker/framework/compute-rocm-dkms-no-npi-hipclang/14599_ubuntu20.04_py3.9_pytorch_rocm6.3_internal_testing_49459b0/
it could have different build issue besides cupy.

**Comment by mawong-amd** [comment]
No objections for this change but I would like to see more justification for the error.

CuPy is not and should not be built by default, even in testing. So I'd like to know if testing is using a container build backend that does not respect build args (i.e. not docker buildkit), or if they're mistakenly building CuPy for some reason.

**Comment by hongxiayang** [comment]
> No objections for this change but I would like to see more justification for the error.
> 
> CuPy is not and should not be built by default, even in testing. So I'd like to know if testing is using a container build backend that does not respect build args (i.e. not docker buildkit), or if they're mistakenly building CuPy for some reason.

the user passed the arguments as below, which i think is ok:
image


It seems to me that it always builds it (https://github.com/ROCm/vllm/blob/4e9830efedb85c86031944df905a3980775bffe0/Dockerfile.rocm#L90) , just based on the flag BUILD_CUPY, it won't copy the build artifacts (whls) when the flag is 0. (https://github.com/ROCm/vllm/blob/4e9830efedb85c86031944df905a3980775bffe0/Dockerfile.rocm#L102). I might be wrong, but please share your explanation.

**Comment by hongxiayang** [comment]
The user is now having a new issue after cupy issue is resolved. This is related to amd-smi build with rocm6.3.
```
18:55:18  Building wheels for collected packages: amdsmi
18:55:18    Building wheel for amdsmi (pyproject.toml): started
18:55:18    Building wheel for amdsmi (pyproject.toml): finished with status 'done'
18:55:18    Created wheel for amdsmi: filename=amdsmi-24.6.3+1a8462f-py3-none-any.whl size=640455 sha256=39872087e10e7cb6869b858b40d7c7047324a786a3449466039d8832bd3ab76e
18:55:18    Stored in directory: /tmp/pip-ephem-wheel-cache-hp8o6u9k/wheels/72/f5/7d/5001c6d6f54c5721bf3548169db6b2e77d20fa0b8066a5f0ac
18:55:18  Successfully built amdsmi
18:55:18  Removing intermediate container 09cb37e3c17f
18:55:18   ---> 966f6da0ae0f
18:55:18  Step 57/101 : FROM scratch AS export_amdsmi
18:55:18   ---> 
18:55:18  Step 58/101 : COPY --from=build_amdsmi /opt/rocm/share/amd_smi/dist/*.whl /
18:55:21  COPY failed: no source files were specified
```

**Comment by mawong-amd** [comment]
> > No objections for this change but I would like to see more justification for the error.
> 
> > 
> 
> > CuPy is not and should not be built by default, even in testing. So I'd like to know if testing is using a container build backend that does not respect build args (i.e. not docker buildkit), or if they're mistakenly building CuPy for some reason.
> 
> 
> 
> the user passed the arguments as below, which i think is ok:
> 
> image
> 
> 
> 
> 
> 
> It seems to me that it always builds it (https://github.com/ROCm/vllm/blob/4e9830efedb85c86031944df905a3980775bffe0/Dockerfile.rocm#L90) , just based on the flag BUILD_CUPY, it won't copy the build artifacts (whls) when the flag is 0. (https://github.com/ROCm/vllm/blob/4e9830efedb85c86031944df905a3980775bffe0/Dockerfile.rocm#L102). I might be wrong, but please share your explanation. 
> 
> 

The Dockerfile logic is admittedly a bit convoluted but here's how it works:

- If `BUILD_CUPY=0`, `export_cupy` is an alias for `export_cupy_0` which in turn is an alias for `scratch`. The `final` layer checks `export_cupy` = `scratch` for any wheels and does not find them as `scratch` is empty. Nothing else happens. In particular, at no time is the `build_cupy` layer accessed, so CuPy wheels are not built.
- If `BUILD_CUPY=1`, `export_cupy` is an alias for `export_cupy_1` which in turn contains CuPy wheels copied over from `build_cupy`, where they were built. The `final` layer checks `export_cupy` = `export_cupy_1` and finds these wheels, installing them.

To summarize, when `BUILD_CUPY=0`, the `build_cupy` stage is not a dependency, so it is never accessed => CuPy wheels are not built.

Older Docker build backends do not support this conditional stage logic and will build all stages, rather than just dependent ones. My next step would be to try the container build after setting the environment variable  `DOCKER_BUILDKIT=1`, to see if this is the issue.

**Review by shajrawi — APPROVED** [review]
